libs/Microsoft.MixedReality.WebRTC.Native/src/interop/global_factory.h

```diff
@@ -16,24 +16,59 @@ enum class ObjectType : int {
 
 /// Global factory wrapper adding thread safety to all global objects, including
 /// the peer connection factory, and on UWP the so-called "WebRTC factory".
+/// Usage pattern:
+///   - Get a RefPtr<> to the singleton instance using |InstancePtr()|. This may
+///   initialize a new instance. This also adds a temporary reference to the
+///   factory to keep it alive until the RefPtr<> goes out of scope.
+///   - Use the factory, for example to get the peer connection factory or to
+///   add/remove some wrapper objects. Whether or not any operation fails, the
+///   factory cannot be destroyed due to the temporary reference, even if no
+///   alive object is present anymore.
+///   - When the RefPtr<> goes out of scope, it releases its temporary reference
+///   to the factory. If that reference is the last one, then the factory
+///   attempts to shutdown, that is goes to check the alive objects, which
+///   constitutes long-term references and will also prevent its destruction.
+///
+/// The delayed call to |TryShutdown()| from releasing the temporary reference
+/// when RefPtr<GlobalFactory> goes out of scope instead of when no object is
+/// alive anymore allows the caller to create a wrapper object and fail
+/// initializing it, and remove it from the global factory but without the
+/// factory destroying itself yet. This works around a problem where the object
+/// removing itself triggers the destruction of the global factory singleton
+/// instance while another thread already acquired a reference to the global
+/// factory but did not yet add any wrapper object so cannot prevent its
+/// destruction, and is left with a dangling reference to a destroyed singleton.
+/// The temporary reference count allows delaying that destruction for as long
+/// as any thread has a reference to the singleton instance. This is not a real
+/// reference count because the global factory is a singleton, and to avoid
+/// circular references. Users must never store a RefPtr<> to the global
+/// factory, but must use a local variable instead to temporarily block
+/// destruction during a single API call.

```

Comment:
s/temporary reference count/temporary reference
(the RefPtr holds a "reference", while the "reference count" is the _number_ of references)


> This is not a real reference count because the global factory is a singleton, and to avoid circular references

Not incredibly clear, I'd rephrase as
> Note that, after any wrapper is added, the factory will be kept alive by the wrapper itself, so it is not ever necessary to store a RefPtr to the factory (in fact, it might cause circular references). As a rule, users should never store...

Reply:
It's indirectly by the size of `alive_objects_`; there is no direct `RefPtr<>` inside the wrappers.